### PR TITLE
Integrate esdt multitransfer

### PR DESCRIPTION
**Pull request by sasurobert**
Integrate esdt multitransfer

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -1153,6 +1152,21 @@ func (sc *scProcessor) isCrossShardESDTTransfer(tx data.TransactionHandler) (str
 		return returnData, true
 	}
 
+	if function == vmcommon.BuiltInFunctionMultiESDTNFTTransfer {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1155]: suggestion to have this functionality embedded in the parser directly?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: go.mod [hunk 2]
```diff
@@ -49,4 +49,6 @@ require (
 
 replace github.com/gogo/protobuf => github.com/ElrondNetwork/protobuf v1.3.2
 
-replace github.com/ElrondNetwork/arwen-wasm-vm/v1_3 v1.3.19 => github.com/ElrondNetwork/arwen-wasm-vm v1.3.20-0.20210709090429-e03405ee6e0b
+replace github.com/ElrondNetwork/arwen-wasm-vm/v1_2 v1.2.26 => github.com/ElrondNetwork/arwen-wasm-vm v1.2.26
+
+replace github.com/ElrondNetwork/arwen-wasm-vm/v1_3 v1.3.23 => github.com/ElrondNetwork/arwen-wasm-vm v1.3.20-0.20210712103947-bff6903b86bf
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply go.mod:54]: yes. for sure. after reviews. and after vmcommon release.

File: process/smartContract/process.go [hunk 3]
```diff
@@ -206,6 +208,12 @@ func NewSmartContractProcessor(args ArgsNewSmartContractProcessor) (*scProcessor
 		incrementSCRNonceInMultiTransferEnableEpoch: args.IncrementSCRNonceInMultiTransferEnableEpoch,
 	}
 
+	var err error
+	sc.esdtTransferParser, err = parsers.NewESDTTransferParser(args.Marshalizer)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:212]: easier to code like this.

File: process/smartContract/process.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/process.go:1155]: this is pretty specific. and used in one place only.

File: process/coordinator/transactionType.go [hunk 4]
```diff
@@ -47,21 +51,30 @@ func NewTxTypeHandler(
...
 		pubkeyConv:             args.PubkeyConverter,
 		shardCoordinator:       args.ShardCoordinator,
 		argumentParser:         args.ArgumentParser,
-		builtInFuncNames:       args.BuiltInFuncNames,
+		builtInFunctions:       args.BuiltInFunctions,
 		relayedTxV2EnableEpoch: args.RelayedTxV2EnableEpoch,
 	}
 
+	var err error
+	tc.esdtTransferParser, err = parsers.NewESDTTransferParser(args.Marshalizer)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/coordinator/transactionType.go:73]: less changes in tests and constructor by doing this :)

File: process/smartContract/process.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:212]: less changes in tests and constructor by doing this :)

File: process/smartContract/builtInFunctions/factory.go [hunk 5]
```diff
@@ -19,6 +19,7 @@ type ArgsCreateBuiltInFunctionContainer struct {
 	Marshalizer          marshal.Marshalizer
 	Accounts             state.AccountsAdapter
 	ShardCoordinator     sharding.Coordinator
+	EpochNotifier        interface{}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/factory.go:22]: done

File: process/smartContract/process.go [hunk 6]
```diff
@@ -1153,6 +1152,21 @@ func (sc *scProcessor) isCrossShardESDTTransfer(tx data.TransactionHandler) (str
...
+			numTransferArgs -= len(parsedTransfer.CallArgs) + 1
+		}
+		if numTransferArgs < 4 {
+			return "", false
+		}
+
+		for i := 0; i < numTransferArgs-1; i++ {
+			returnData += hex.EncodeToString(args[i]) + "@"
+		}
+		returnData += hex.EncodeToString(args[numTransferArgs-1])
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:1167]: done. missed the return

File: process/smartContract/process.go [hunk 7]
```diff
@@ -1153,6 +1152,21 @@ func (sc *scProcessor) isCrossShardESDTTransfer(tx data.TransactionHandler) (str
 		return returnData, true
 	}
 
+	if function == vmcommon.BuiltInFunctionMultiESDTNFTTransfer {
+		numTransferArgs := len(args)
+		if len(parsedTransfer.CallFunction) > 0 {
+			numTransferArgs -= len(parsedTransfer.CallArgs) + 1
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:1158]: numTransferArgs = numTransferArgs - len(parsedTransfer.CallArgs) - 1

[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/smartContract/process.go:1158]: ok

**Review by SebastianMarian — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by SebastianMarian — DISMISSED** [review]

**Review by iulianpascalau — CHANGES_REQUESTED** [review]
forgotten set of the `ESDTMultiTransferEnableEpoch` in genesis/process/shardGenesisBlockCreator.go L279.

+ the change in process/coordinator/transactionType.go (see comment above)

File: process/smartContract/process.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/smartContract/process.go:212]: Usually disagree with this pattern.

File: process/coordinator/transactionType.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/coordinator/transactionType.go:73]: Please read again the comment. Please move this on L60.

**Review by iulianpascalau — APPROVED** [review]

**Review by liviuancas-elrond — APPROVED** [review]
tests ok

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**